### DISCUSSION TITLE
My Mocha tests fail when trying to use "Ref" in an ESM project

### DISCUSSION THREAD

**Discussion by neilenns** [typegoose/typegoose#875] [Q&A] [answered]
I have a project that uses modules. The project codebase builds fine and runs perfectly, but Mocha fails to run tests with this error:

> SyntaxError: Named export 'Ref' not found. The requested module '@typegoose/typegoose' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

> import pkg from '@typegoose/typegoose';
const { modelOptions, prop, plugin, getModelForClass, Ref, } = pkg;

It doesn't matter if I use the alternate method of importing, I still get an error about Ref not being found. This is the only Typegoose piece I have issues with. If I don't use Ref Mocha has no issues testing any of the other code that depends on Typegoose.

tsconfig.json has the following set:

```
"module": "esnext"
"moduleResolution": "node"
```

package.json has the following set:

```
"type": "module",
```

.mocharc.jsonc has the following set:

```
  "node-option": ["experimental-specifier-resolution=node", "loader=ts-node/esm"],
```

Does anyone know how to resolve this issue? Since the code runs fine outside of Mocha I assume it is some sort of configuration thing I'm missing, but I can't figure out what it might be.

**Comment by hasezoey** [comment]
this issue seems to be that somehow the *type* `Ref` gets compiled into the output, and is so interpreted as a *runtime value*, which does not exist, maybe check if you somehow have `isolatedModules` set to `true` or something similar

using `const { modelOptions, prop, plugin, getModelForClass, Ref, } = pkg;` means that you want to import *value* `Ref`, not *type* `Ref` (meaning this will never work)

**Reply by neilenns** [reply]
Thanks for the quick response!

I took another pass through the project and `isolatedModules` isn't being set anywhere in the server code. It is set in the client code, but that's a separate folder in the monorepo.

For completeness I tried putting `isolatedModules` in the tsconfig.json for the server and setting it explicitly to `false` and that didn't change anything. The mocha run still gives the same error :(

**Reply by neilenns** [reply]
Here's the complete set of compiler options for the server:

```
    "target": "es2020" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
    "experimentalDecorators": true /* Enable experimental support for legacy experimental decorators. */,
    "emitDecoratorMetadata": true /* Emit design-type metadata for decorated declarations in source files. */,
    "module": "esnext" /* Specify what module code is generated. */,
    "moduleResolution": "node" /* Specify how TypeScript looks up a file from a given module specifier. */,
    "typeRoots": ["./types"] /* Specify multiple folders that act like './node_modules/@types'. */,
    "resolveJsonModule": true /* Enable importing .json files. */,
    "sourceMap": true /* Create source map files for emitted JavaScript files. */,
    "outDir": "./dist" /* Specify an output folder for all emitted files. */,
    "esModuleInterop": true /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables 'allowSyntheticDefaultImports' for type compatibility. */,
    "forceConsistentCasingInFileNames": true /* Ensure that casing is correct in imports. */,
    "strict": true /* Enable all strict type-checking options. */,
    "skipLibCheck": true /* Skip type checking all .d.ts files. */
```

**Reply by hasezoey** [reply]
i dont know, but maybe `skipLibCheck` may be related?

aside from that, i dont think i will be able to help debug this without a reproduction repository

**Reply by neilenns** [reply]
Ok thanks. I'll try skiplibcheck and if that doesn't work see if I can set up a minimal repo.

**Comment by neilenns** [comment]
Attached zip file has a minimal repo. See the README inside, but basically:

1. Run `npm install`
2. Run `npm run build`. This will succeed.
3. Run `npm run start`. This will succeed.
4. Run `npm run test`. This will fail.

[minimal-repo.zip](https://github.com/typegoose/typegoose/files/12317259/minimal-repo.zip)

**Reply by hasezoey** [reply, marked as answer]
your reproduction repository does not actually contain any tests, so i cannot say any further than the first issue is that you have `ts-node: transpileOnly: true` set, which does no type-checking, which seems to be required for combined type and value imports

maybe somewhat related to https://github.com/TypeStrong/ts-loader/issues/751

workaround without removing that option would be to split type and value imports, like
```ts
import { getModelForClass } from "...";
import type { Ref } from "...";
```

though this might become annoying to do and check manually (because things like vscode will always try to combine them)

**Reply by neilenns** [reply]
`transpileOnly: true` was the problem. Thank you!!!

> your reproduction repository does not actually contain any tests

There's a setup fixture that was sufficient to reproduce the issue without actually running a specific test.
